File: 1001-1100/1057-Campus Bikes/1057-Campus Bikes.cpp
class Solution {
public:
    vector<int> assignBikes(vector<vector<int>>& workers, vector<vector<int>>& bikes) {
        vector<vector<pair<int, int>>> distance(2001);
        for (int i = 0; i < workers.size(); ++i) {
            for (int j = 0; j < bikes.size(); ++j) {
                int d = abs(workers[i][0] - bikes[j][0]) + abs(workers[i][1] - bikes[j][1]);
                distance[d].push_back({i, j});
            }
        }
        
        vector<int> result(workers.size(), -1);
        vector<bool> assigned(bikes.size());
        for (int d = 1; d <= 2000; ++d) {
            for (int i = 0; i < distance[d].size(); ++i) {
                int wi = distance[d][i].first;
                int bi = distance[d][i].second;
                if (result[wi] == -1 && !assigned[bi]) {
                    result[wi] = bi;
                    assigned[bi] = true;
                }
            }
        }
        
        return result;
    }
};

File: 1001-1100/1057-Campus Bikes/1057-Campus Bikes.py
import collections
class Solution:
    def assignBikes(self, workers: List[List[int]], bikes: List[List[int]]) -> List[int]:
        distance = collections.defaultdict(list)
        for i, w in enumerate(workers):
            for j, b in enumerate(bikes):
                distance[abs(w[0] - b[0]) + abs(w[1] - b[1])].append((i, j))

        result = [-1] * len(workers)
        assigned = [False] * len(bikes)
        for d in sorted(distance):
            for wi, bi in distance[d]:
                if result[wi] == -1 and not assigned[bi]:
                    result[wi] = bi
                    assigned[bi] = True

        return result
